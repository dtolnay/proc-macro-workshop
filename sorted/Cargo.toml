[package]
name = "sorted"
version = "0.0.0"
autotests = false
edition = "2021"
publish = false

[lib]
proc-macro = true

[[test]]
name = "tests"
path = "tests/progress.rs"

[dev-dependencies]
trybuild = { version = "1.0.49", features = ["diff"] }

[dependencies]
proc-macro2 = "1.0"
quote = "1.0"
syn = { version = "2.0", features = ["full", "visit-mut"] }
