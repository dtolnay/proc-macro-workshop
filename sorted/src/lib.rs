mod check;
mod compare;
mod emit;
mod format;
mod parse;
mod visit;

use proc_macro::TokenStream;
use quote::quote;
use syn::{parse_macro_input, ItemFn};

use crate::emit::emit;
use crate::parse::{Input, Nothing};

#[proc_macro_attribute]
pub fn sorted(args: TokenStream, input: TokenStream) -> TokenStream {
    let original = input.clone();

    let _ = parse_macro_input!(args as Nothing);
    let input = parse_macro_input!(input as Input);
    let kind = input.kind();

    match check::sorted(input) {
        Ok(()) => original,
        Err(err) => emit(err, kind, original),
    }
}

#[proc_macro_attribute]
pub fn check(args: TokenStream, input: TokenStream) -> TokenStream {
    let _ = parse_macro_input!(args as Nothing);
    let mut input = parse_macro_input!(input as ItemFn);

    visit::check(&mut input);

    TokenStream::from(quote!(#input))
}
