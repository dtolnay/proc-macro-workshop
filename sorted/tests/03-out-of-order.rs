// At this point we have an enum and we need to check whether the variants
// appear in sorted order!
//
// When your implementation notices a variant that compares lexicographically
// less than one of the earlier variants, you'll want to construct a syn::Error
// that gets converted to TokenStream by the already existing code that handled
// this conversion during the previous test case.
//
// The "span" of your error, which determines where the compiler places the
// resulting error message, will be the span of whichever variant name that is
// not in the right place. Ideally the error message should also identify which
// other variant the user needs to move this one to be in front of.

use sorted::sorted;

#[sorted]
pub enum Error {
    ThatFailed,
    ThisFailed,
    SomethingFailed,
    WhoKnowsWhatFailed,
}

fn main() {}
