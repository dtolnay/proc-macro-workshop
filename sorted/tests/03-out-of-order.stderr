error: SomethingFailed should sort before ThatFailed
   |
20 |     SomethingFailed,
   |     ^^^^^^^^^^^^^^^
